### PR TITLE
Weighted centroid

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---
There are a couple of new features here, but mainly this is to include the functionality to generate centroids based on weighting on lower zones (e.g. emp, pop).

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`
- [ ] Have any new features been properly integrated with the API (CLI and GUI)?

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -353,6 +354,12 @@ def fixture_weighted_trans(weighted_config) -> pd.DataFrame:
     return trans
 
 
+@pytest.fixture(name="centroids", scope="session")
+def fixture_centroids(weighted_config) -> pd.DataFrame:
+    trans = zone_translation.ZoneTranslation(weighted_config).weighted_centroids()
+    return trans
```
Comment (MattBuckley-TfN) [inline tests/conftest.py]: This would probably make more sense in the actual test instead of as a fixture because this is the bit of code we're actually testing, then this fixture can be removed.

File: tests/test_zone_translation.py [hunk 2]
```diff
@@ -265,3 +279,7 @@ def test_output(self, trans_str: str, expected_str: str, request):
         df_2 = expected.groupby(["zone_1_id", "zone_2_id"]).sum()
         df_2.sort_index(inplace=True)
         pd.testing.assert_frame_equal(df_1, df_2)
+
+    def test_centroids(self, centroids, expected_centroids):
+        rounded = centroids.round(2)
+        pd.testing.assert_frame_equal(rounded, expected_centroids)
```
Comment (MattBuckley-TfN) [inline tests/test_zone_translation.py]: This should run the code actual being tested instead of having the centroids fixture run it, e.g.
```suggestion
    def test_centroids(self, weighted_config, expected_centroids):
        centroids = zone_translation.ZoneTranslation(weighted_config).weighted_centroids()
        rounded = centroids.round(2)
        pd.testing.assert_frame_equal(rounded, expected_centroids)
```

Also, it could do with type annotations and a docstring

File: tests/test_zone_translation.py [hunk 3]
```diff
@@ -106,6 +106,20 @@ def fixture_expected_points() -> pd.DataFrame:
     return output
 
 
+@pytest.fixture(name="expected_centroids", scope="class")
+def fixture_expected_centroids() -> pd.DataFrame:
```
Comment (MattBuckley-TfN) [inline tests/test_zone_translation.py:110]: Please could you add a docstring explaining that this fixtures is the expected results for the test centroids test below, just to make it more obvious what the results are for

File: src/scripts/comp_zone.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
```
Comment (MattBuckley-TfN) [inline src/scripts/comp_zone.py:1]: What's this file for and why is it in a folder called early? If it's a prototype / helper script I tend to put these in a `src\scripts` folder so they're outside the package but still in the source code folder.

File: src/scripts/itn_filter.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import geopandas as gpd
```
Comment (MattBuckley-TfN) [inline src/scripts/itn_filter.py:3]: What's this file for and why is it in a folder called early? If it's a prototype / helper script I tend to put these in a `src\scripts` folder so they're outside the package but still in the source code folder.

File: examples/weighted_run.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+params = cs.ZoningTranslationInputs(
+    zone_1=zone_1,
+    zone_2=zone_2,
+    lower_zoning=lower,
+    method="pop",
+    rounding=True,
+)
+trans_class = cs.ZoneTranslation(params)
+trans_df = trans.weighted_translation()
```
Comment (MattBuckley-TfN) [inline examples/weighted_run.py]: Typo in variable name
```suggestion
trans_df = trans_class.weighted_translation()
```

File: src/caf/space/__main__.py [hunk 7]
```diff
@@ -41,12 +41,17 @@ def main():
...
                 trans.weighted_translation().to_csv(
                     args.out_path
                     / f"{config.zone_1.name}_{config.zone_2.name}_{config.method}.csv",
                     index=False,
                 )
+            elif args.mode == "centroid":
+                trans.weighted_centroids().to_csv(
+                    args.out_path
+                    / f"{config.zone_1.name}_{config.lower_zoning.name}_{config.method}.csv",
+                )
```
Comment (MattBuckley-TfN) [inline src/caf/space/__main__.py:54]: Add an else statement to raise an error for any unexpected modes
```suggestion
                )
            else:
                raise ValueError(f"invalid mode: {args.mode}")
```

File: src/caf/space/inputs.py [hunk 8]
```diff
@@ -99,27 +99,28 @@ class LowerZoneSystemInfo(ZoneSystemInfo):
...
         to the cache and is important for logging. If you don't know this you
         should consider whether your weighting data is appropriate.
     """
 
-    weight_data: Path
+    weight_data: Optional[Path] = None
     data_col: str
-    weight_id_col: str
+    weight_id_col: Optional[str] = None
     weight_data_year: int
```
Comment (MattBuckley-TfN) [inline src/caf/space/inputs.py:124]: Is this ordering allowed? I'd expect it to raise an error about parameters with defaults being before mandatory ones?

File: src/caf/space/utils.py [hunk 9]
```diff
@@ -120,3 +121,11 @@ def points_update(
     points.set_index(id_col, inplace=True)
     points.drop(list(matches[matches_id]), axis=0, inplace=True)
     return points.reset_index()
+
+
+def set_crs(gdf: gpd.GeoDataFrame, name: str):
```
Comment (MattBuckley-TfN) [inline src/caf/space/utils.py]: Missing return type
```suggestion
def set_crs(gdf: gpd.GeoDataFrame, name: str) -> gpd.GeoDataFrame:
```

File: src/caf/space/utils.py [hunk 10]
```diff
@@ -120,3 +121,11 @@ def points_update(
...
+    """Set crs of gdf if missing."""
+    if gdf.crs is None:
+        warnings.warn(f"Zone {name} has no CRS, setting crs to EPSG:27700.", stacklevel=2)
+        gdf.set_crs("EPSG:27700")
```
Comment (MattBuckley-TfN) [inline src/caf/space/utils.py]: This isn't done inplace so should be set to gdf
```suggestion
    """Set crs of gdf if missing."""
    if gdf.crs is None:
        warnings.warn(f"Zone {name} has no CRS, setting crs to EPSG:27700.", stacklevel=2)
        gdf = gdf.set_crs("EPSG:27700")
```

File: src/caf/space/zone_translation.py [hunk 11]
```diff
@@ -48,34 +48,40 @@ class ZoneTranslation:
...
         self.cache_path = params.cache_path
-        if params.lower_zoning:
+        self.zone_1 = params.zone_1
+        if params.zone_2 is not None:
+            self.zone_2 = params.zone_2
+        if params.lower_zoning is not None:
             self.lower_zoning = params.lower_zoning
-        if params.method:
+        if params.method is not None:
             self.method = params.method
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py:62]: The instance variables should probably be set to the parameters even if they're None because otherwise they're undefined which can cause confusing bugs. Tbh I think this would be fixed by splitting out the translation functionality in the refactor.

File: src/caf/space/zone_translation.py [hunk 12]
```diff
@@ -48,34 +48,40 @@ class ZoneTranslation:
...
         self.slither_tolerance = params.sliver_tolerance
         self.rounding = params.rounding
         self.filter_slithers = params.filter_slivers
         self.point_handling = params.point_handling
         self.point_tolerance = params.point_tolerance
         self.run_date = params.run_date
-        sorted_names = sorted([params.zone_1.name, params.zone_2.name])
-        self.names = (sorted_names[0], sorted_names[1])
         self.logger = logging.getLogger("SPACE")
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py:69]: Probably one for the refactor but it would be better to define a module level logger based on `__name__` because this "SPACE" logger isn't being controlled by the log setup in `__main__`.

File: src/caf/space/zone_translation.py [hunk 13]
```diff
@@ -48,34 +48,40 @@ class ZoneTranslation:
...
+        self.names = (sorted_names[0], sorted_names[1])
         self.handler = logging.FileHandler(
             self.cache_path / f"{self.names[0]}_{self.names[1]}.log", mode="w"
         )
+
         self.handler.setFormatter(
             logging.Formatter("%(asctime)s [%(name)-20.20s] [%(levelname)-8.8s]  %(message)s")
         )
         self.logger.addHandler(self.handler)
         self.logger.setLevel(logging.INFO)
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py:89]: It's not best practise to setup logging within the functionality of the package, it's better to set it up as part of the front-end then if anyone imports this module to use the class in their own code they're able to manage the logging, whereas right now this takes control away from the user.

If we want to track info within the cache it might be better to just output all the parameters (plus some tool and system information) using caf.toolkit's `write_metadata` function.

File: src/caf/space/zone_translation.py [hunk 14]
```diff
@@ -48,34 +48,40 @@ class ZoneTranslation:
...
 
-    def spatial_translation(self) -> pd.DataFrame:
+    def spatial_translation(self, return_gdf: bool = False) -> pd.DataFrame:
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py]: Another one for the refactor, I think it's probably better to just have a separate method which performs the translation and returns a GeoDataFrame instead of having a parameter because it makes the type annotations much cleaner.

For now I think fixing the return type should be enough
```suggestion
    def spatial_translation(self, return_gdf: bool = False) -> pd.DataFrame | gpd.GeoDataFrame:
```

File: src/caf/space/zone_translation.py [hunk 15]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
...
+        out_path = self.cache_path / f"{self.names[0]}_{self.names[1]}"
+        out_path.mkdir(exist_ok=True, parents=False)
+        out_name = f"{self.names[0]}_to_{self.names[1]}_weighted_centroids"
+        grouped[["x", "y"]].to_csv(out_path / f"{out_name}.csv")
+        self.params.save_yaml(out_path / f"{out_name}.yml")
+        return grouped[["x", "y"]]
+
+    def _slithers_and_rounding(
+        self, translation: pd.DataFrame, return_gdf: bool = False
+    ) -> pd.DataFrame:
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py]: Incorrect return type
```suggestion
    def _slithers_and_rounding(
        self, translation: pd.DataFrame, return_gdf: bool = False
    ) -> pd.DataFrame | gpd.GeoDataFrame:
```

File: src/caf/space/zone_translation.py [hunk 16]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
         self.params.save_yaml(out_path / f"{out_name}.yml")
         return weighted_translation
 
-    def _slithers_and_rounding(self, translation: pd.DataFrame) -> pd.DataFrame:
+    def weighted_centroids(self):
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py]: Missing return type
```suggestion
    def weighted_centroids(self) -> pd.DataFrame:
```

File: src/caf/space/zone_translation.py [hunk 17]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
...
+        """
+        Create centroids based on weighting data.
+
+        Returns
+        -------
+        pd.DataFrame: A dataframe of centroids, with columns for 'x' and 'y'.
+        The zone ids will be in the index.
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py]: Worth including CRS of x / y in the return description

File: src/caf/space/zone_translation.py [hunk 18]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
...
+            zone_2=self.lower_zoning._lower_to_higher(),
+            cache_path=self.cache_path,
+            sliver_tolerance=self.slither_tolerance,
+            rounding=self.rounding,
+            filter_slivers=self.filter_slithers,
+        )
+        lookup = ZoneTranslation(trans_conf).spatial_translation(return_gdf=True)
+        lower_weight = pd.read_csv(
+            self.lower_zoning.weight_data, index_col=self.lower_zoning.weight_id_col
+        )
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py:250]: Add `usecols` to make sure the right columns are there and nothing unnecessary is being read
```suggestion
        lower_weight = pd.read_csv(
            self.lower_zoning.weight_data,
            index_col=self.lower_zoning.weight_id_col,
            usecols=[self.lower_zoning.weight_id_col, self.lower_zoning.data_col],
        )
```

File: src/caf/space/zone_translation.py [hunk 19]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
...
+        lower_weight.index.name = f"{self.lower_zoning.name}_id"
+        cent = lookup.join(lower_weight)
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py:252]: Is this going to have the same error that James found if column names happen to be the same? If so it's probably worth adding a check or suffixes. The columns could be renamed to something generic when loaded to avoid this.

File: src/caf/space/zone_translation.py [hunk 20]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
...
+        cent[self.lower_zoning.data_col] *= cent[
+            f"{self.lower_zoning.name}_to_{self.zone_1.name}"
+        ]
+        cent["x_weight"] = cent.centroid.x * cent[self.lower_zoning.data_col]
+        cent["y_weight"] = cent.centroid.y * cent[self.lower_zoning.data_col]
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py]: Extremely unlikely but if data_col is called "x_weight" this will cause a problem with the calculation but won't raise an error so it might be worth renaming any columns when loading the weight data to avoid name clashes

File: src/caf/space/zone_translation.py [hunk 21]
```diff
@@ -197,7 +211,49 @@ def weighted_translation(self) -> pd.DataFrame:
...
+        grouped = cent.groupby(f"{self.zone_1.name}_id")[
+            [self.lower_zoning.data_col, "x_weight", "y_weight"]
+        ].sum()
+        grouped["x"] = grouped["x_weight"] / grouped[self.lower_zoning.data_col]
+        grouped["y"] = grouped["y_weight"] / grouped[self.lower_zoning.data_col]
```
Comment (MattBuckley-TfN) [inline src/caf/space/zone_translation.py]: Future reference I think this could be done with numpy's weighted average function (`np.average`)

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Logic looks good, just some comments tidying up and adding docs / type annotations.

Also, one thing I noticed when `caf.space.inputs.SpaceArguments` is used in `__main__` is that VS code (and possibly linters) isn't picking it up correctly I think because the `inputs` module isn't imported directly in `__init__`. It doesn't actually cause a Python error though I think because of `from caf.space.inputs import ...`, but I think `from caf.space import inputs` should be added to `__init__` anyway to make sure IDEs and linters all work correctly.

File: src/caf/space/inputs.py (hunk 8 above)
Comment (isaac-tfn) [reply src/caf/space/inputs.py:124]: It's allowed in BaseConfig classes (I think)

File: src/caf/space/zone_translation.py (hunk 12 above)
Comment (isaac-tfn) [reply src/caf/space/zone_translation.py:69]: There is a global LOG so I don't know why I've done this really

File: src/caf/space/zone_translation.py (hunk 19 above)
Comment (isaac-tfn) [reply src/caf/space/zone_translation.py:252]: I think this is resolved by the suggestion applied above?

**Review by MattBuckley-TfN — APPROVED** [review]
LGTM

**Comment by MattBuckley-TfN** [comment]
> LGTM

Actually, @isaac-tfn please could you run black

**Comment by MattBuckley-TfN** [comment]
Thanks, looks good to merge @isaac-tfn
